cpu/atmega2560/periph/gpio.c

```diff
@@ -29,47 +31,440 @@
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            GPIO_1_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            GPIO_2_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            GPIO_3_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            GPIO_4_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            GPIO_5_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+    };
+
+    return 0;
 }
 
 int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t cb, void *arg)
 {
-    return -1;
+    int ret_code;
+    uint8_t pin_B;
+    uint8_t pin;
+
+    ret_code = gpio_init_in(dev, pushpull);
+    if (ret_code < 0)
+    {
+        return ret_code;
+    }
+
+    pin = gpio_pin_map[dev];
+
+    /* clear global interrupt flag */
+    cli();
+
+    switch(dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+    };
+
+    pin_B = (pin > GPIO_3 && pin == GPIO_4) ? 0 : 1;
+
+    /* configure the flank */
+    switch (flank) {
+        case GPIO_RISING:
+            if (pin < 4)
+            {
+                EICRA |= (3 << pin * 2);
+            }
+            else
+            {
+                EICRB |= (3 << pin_B * 2);
+            }
+            break;
+        case GPIO_FALLING:
+            if (pin < 4)
+            {
+                EICRA |= (2 << pin * 2);
+            }
+            else
+            {
+                EICRB |= (2 << pin_B * 2);
+            }
+            break;
+        case GPIO_BOTH:
+            if (pin < 4)
+            {
+                EICRA |= (1 << pin * 2);
+            }
+            else
+            {
+                EICRB |= (1 << pin_B * 2);
+            }
+            break;
+    };
+
+    /* set callback */
+    config[dev].cb = cb;
+    config[dev].arg = arg;
+
+    /* set global interrupt flag */
+    sei();
+
+    return 0;
+}
+
+static inline void irq_handler(gpio_t dev)
+{
+    config[dev].cb(config[dev].arg);
+    /*
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+    */
+}
+
+#if GPIO_0_EN
+ISR(INT0_vect, ISR_BLOCK)
```

Comment:
I think you also need to abstract `INT0_vect` in the periph_conf file. Imagine The pin of `GPIO_0` changes. Would this still be the correct interrupt? I don't know exactly. So don't change just in cause of this comment :-) !